src/databricks/sql/auth/oauth.py

```diff
@@ -0,0 +1,286 @@
+
+    with HTTPServer(("", port), SingleRequestHandler) as httpd:
+        webbrowser.open_new(auth_req_uri)
+        click.echo("Listening for OAuth authorization callback at {uri}"
+                   .format(uri=redirect_url))
+        httpd.handle_request()
+
+    if not global_request_path:
+        error_and_quit("No path parameters were returned to the callback at {uri}"
+                       .format(uri=redirect_url))
+    # This is a kludge because the parsing library expects https callbacks
+    # We should probably set it up using https
+    full_redirect_url = "https://localhost:{port}/{path}".format(
+        port=port, path=global_request_path)
+    try:
+        authorization_code_response = \
+            client.parse_request_uri_response(full_redirect_url, state=state)
+    except OAuth2Error as err:
+        error_and_quit("OAuth Token Request error {error}".format(error=err.description))
+    return authorization_code_response
+
+
+def send_auth_code_token_request(client, token_request_url, redirect_url, code, verifier):
+    token_request_body = client.prepare_request_body(code=code, redirect_uri=redirect_url)
+    data = "{body}&code_verifier={verifier}".format(body=token_request_body, verifier=verifier)
+    return send_token_request(token_request_url, data)
+
+
+def send_token_request(token_request_url, data):
+    headers = {
+        "Accept": "application/json",
+        "Content-Type": "application/x-www-form-urlencoded"
+    }
+    response = requests.request(method="POST", url=token_request_url, data=data, headers=headers)
+    oauth_response = json.loads(response.text)
+    return oauth_response
+
+
+def send_refresh_token_request(hostname, client_id, refresh_token):
+    idp_url = get_idp_url(hostname)
+    oauth_config = fetch_well_known_config(idp_url)
+    token_request_url = oauth_config['token_endpoint']
+    client = get_client(client_id=client_id)
+    token_request_body = client.prepare_refresh_body(
+        refresh_token=refresh_token, client_id=client.client_id)
+    return send_token_request(token_request_url, token_request_body)
+
+
+def get_tokens_from_response(oauth_response):
+    access_token = oauth_response['access_token']
+    refresh_token = oauth_response['refresh_token'] if 'refresh_token' in oauth_response else None
+    return access_token, refresh_token
+
+
+def check_and_refresh_access_token(hostname, access_token, refresh_token):
+    now = datetime.now(tz=UTC)
+    # If we can't decode an expiration time, this will be expired by default.
+    expiration_time = now
+    try:
+        # This token has already been verified and we are just parsing it.
+        # If it has been tampered with, it will be rejected on the server side.
+        # This avoids having to fetch the public key from the issuer and perform
+        # an unnecessary signature verification.
+        decoded = jwt.decode(access_token, options={"verify_signature": False})
```

Comment:
```suggestion
        decoded = json.loads(base64.standard_b64decode(access_token.split(".")[1]))
```

please rewrite it with something like that - don't depend on `PyJWT`, we don't need it here. jwt tokens are just dot-separated base-64-encoded jsons - https://jwt.io/

Reply:
Thanks @nfx this was very helpful.